apollo-router/src/plugins/telemetry/mod.rs

```diff
@@ -1541,6 +1655,128 @@ impl Telemetry {
     }
 }
 
+#[derive(Debug, Clone)]
+struct CacheAttributes {
+    subgraph_name: Arc<String>,
+    headers: http::HeaderMap,
+    hashed_query: Arc<String>,
+    // Typename + hashed_representation
+    representations: Vec<(Arc<String>, Value)>,
+}
+
+#[derive(Debug, Hash, Clone)]
+struct CacheKey {
+    representation: Value,
+    typename: Arc<String>,
+    query: Arc<String>,
+    subgraph_name: Arc<String>,
+    hashed_headers: Arc<String>,
+}
+
+// Get typename and hashed representation for each representations in the subgraph query
+fn extract_cache_attributes(
+    representations: &[Value],
+) -> Result<Vec<(Arc<String>, Value)>, BoxError> {
+    let mut res = Vec::new();
+    for representation in representations {
+        let opt_type = representation
+            .as_object()
+            .and_then(|o| o.get(TYPENAME))
+            .ok_or("missing __typename in representation")?;
+        let typename = opt_type.as_str().unwrap_or("");
+
+        res.push((Arc::new(typename.to_string()), representation.clone()));
+    }
+    Ok(res)
+}
+
+struct CacheCounter {
+    potential_cache_entries: CountMinSketch<CacheKey, usize>,
+    // By subgraph
+    nb_requests: HashMap<Arc<String>, usize>,
+    // By subgraph and typename
+    potential_cache_hit: HashMap<(Arc<String>, Arc<String>), usize>,
+    created_at: Instant,
+    ttl: Duration,
+}
+
+impl CacheCounter {
+    fn new(ttl: Duration) -> Self {
+        Self {
+            potential_cache_entries: CountMinSketch::new(
+                0.01, // probability calculate k_num
+                0.01, // error tolerance to calculate width
+                (),
+            ),
+            potential_cache_hit: HashMap::new(),
+            nb_requests: HashMap::new(),
+            created_at: Instant::now(),
+            ttl,
+        }
+    }
+
+    fn record(
+        &mut self,
+        query: Arc<String>,
+        subgraph_name: Arc<String>,
+        hashed_headers: Arc<String>,
+        representations: Vec<(Arc<String>, Value)>,
+    ) {
+        if self.created_at.elapsed() >= self.ttl {
+            self.clear();
+        }
+
+        // I record the total number of requests (asking for _entities) to a specific subgraph
+        // in order to know the frequency of potential cache hit we could have for a subgraph
+        let nb_requests = *self
+            .nb_requests
+            .entry(subgraph_name.clone())
+            .and_modify(|e| *e += 1)
+            .or_insert(1);
+
+        for (typename, representation) in representations {
+            let cache_hit = self.potential_cache_entries.push(
+                &CacheKey {
+                    representation,
+                    typename: typename.clone(),
+                    query: query.clone(),
+                    subgraph_name: subgraph_name.clone(),
+                    hashed_headers: hashed_headers.clone(),
+                },
+                &1,
+            ) > 1;
```

Comment:
since we are now only caring about having seen an entity, not how many time we've seen it, we should switch back to a bloom filter. The `streaming-algorithms` crate is nice, but has not been updated in 3 years. [bloomfilter](https://crates.io/crates/bloomfilter) should be ok if we accept the ISC license